all-sdk/autotuning.md

```diff
@@ -0,0 +1,452 @@
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots (in use or not) for that task
+  type at all times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For
+  example, in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available. In languages with explicit cancel mechanisms, this should be cancellable and
+  /// return a boolean indicating whether a slot was actually obtained or not. In Rust, the future can simply
+  /// be dropped if the reservation is no longer desired.
+  async fn reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>);
```

Comment:
This is what I originally had, and is what I have in Rust because internally I store a struct that auto-releases on drop.

As for user data though, if we're saying elsewhere we'd prefer to not have something until there's a demonstrated need for it, I think that's probably even more applicable here. I don't know what the immediate use case is for this, but I do have known use cases for the `used_slots` on context.

Reply:
Any language with real destructors, yeah, definitely.

Reply:
> As for user data though, if we're saying elsewhere we'd prefer to not have something until there's a demonstrated need for it, I think that's probably even more applicable here

Less applicable, because this is future proofing. We cannot add a return type to this easily later. The immediate use case would be like etcd lease info. But just an opaque object that we give back to them is nice.

Reply:
I can add it back. I'm just a little worried about how extensive of changes that might result in in Go/Java where there is no permit object like in Core. We'd have to track it everywhere like Core does, which ultimately is nice, but could be quite a lot of code churn.

I think any kind of uniquely identifying a particular slot/permit would have the same consequence. It's useful though, in the abstract.

@Quinn-With-Two-Ns what do you think?